### PR TITLE
feat(plugins): #16 - adding pyaqueduct access to plugins

### PR DESCRIPTION
**Pull request by str-anger**
Implementation of 2 calls:
1. Listing plugins
2. Executing a particular plugin by name

### CODE REVIEW

**Review by gabrielg42 — APPROVED** [review]
Thanks @str-anger !

File: pyaqueduct/api.py [hunk 1]
```diff
@@ -129,3 +130,16 @@ def find_experiments(
...
             for experiment in experiments
         ]
+
+    @validate_call
+    def get_plugins(self) -> List[Plugin]:
+        """Gets the current fresh plugin list from the server.
+        Plugin list may change without server restart.
+
+        Returns:
+            list of plugin objects.
```
Comment (samiralavi) [inline pyaqueduct/api.py:140]: "L" instead of "l" for list.

File: pyaqueduct/client/client.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 from tqdm import tqdm
 
 from pyaqueduct.client.types import ExperimentData, ExperimentsInfo, TagsData
+from pyaqueduct.client.plugin_types import PluginData, PluginExecutionResultData
```
Comment (samiralavi) [inline pyaqueduct/client/client.py:18]: If you have made this decision to separate the types in different modules, please refactor the experiment types to be collected in `experiment_types` module.

File: pyaqueduct/client/client.py [hunk 3]
```diff
@@ -435,3 +438,67 @@ def download_file(self, experiment_uuid: UUID, file_name: str, destination_dir:
...
+        return plugins
+
+    def execute_plugin_function(
+            self, plugin: str, function: str, params: Dict[str, Any]) -> PluginExecutionResultData:
+        """ Executes plugin function on a server.
+
+        Args:
+            plugin: plugin name.
+            function: function name within a plugin.
+            params (Dict[str, Any]): dictionary with parameters passed to a plugin.
```
Comment (samiralavi) [inline pyaqueduct/client/client.py:472]: The type name can be omitted.

File: pyaqueduct/client/client.py [hunk 4]
```diff
@@ -435,3 +438,67 @@ def download_file(self, experiment_uuid: UUID, file_name: str, destination_dir:
...
+
+        Raises:
+            RemoteOperationError: Communication error.
+
+        Returns:
+            PluginExecutionResultData:
```
Comment (samiralavi) [inline pyaqueduct/client/client.py:478]: Type name can be omitted to be coherent with the rest of docstrings.

File: pyaqueduct/plugin.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    parameters: List[PluginParameter]
+    plugin: Optional[Plugin] = None
+
+    _client: Optional[AqueductClient] = None
+
+    # TODO: may require removal from autodocs
+    # if used together with sphinx
+    # see: https://stackoverflow.com/q/28224554
+    def set_client(self, client: AqueductClient):
```
Comment (samiralavi) [inline pyaqueduct/plugin.py:33]: You can override the initialiser similar to the Experiment module.

File: pyaqueduct/plugin.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+class PluginExecutionResult(PluginExecutionResultData):
+    """Class representing result of plugin execution"""
+
+
+class PluginParameter(PluginParameterData):
+    """Plugin parameter class"""
+
+
+class PluginFunction(PluginFunctionData):
```
Comment (samiralavi) [inline pyaqueduct/plugin.py:22]: This inheritance is going to be problematic. You are inheriting from a data class whose purpose is only data transfer. Instead, assign a member variable of this type. You can look at the `Experiment` module.

File: pyaqueduct/plugin.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        Args:
+            parameters: dictionary of parameters to pass to a plugin.
+
+        Raises:
+            AttributeError: client connection should be initialised before this call.
+
+        Returns:
+            result of plugin execution on server. `returnCode==0` corresponds to success.
+        """
+        if self._client is None:
```
Comment (samiralavi) [inline pyaqueduct/plugin.py:49]: You didn't need to catch this exception if you initialised the object with all the required items.

**Review by samiralavi — APPROVED** [review]
Great work @str-anger, it is nice you got the class design so well. I have put some comments for improvement.

File: pyaqueduct/plugin.py
Comment (samiralavi) [inline pyaqueduct/plugin.py:1]: It looks neater now.

**Review by samiralavi — APPROVED** [review]
I looked again at the changes and it is a lot neater now. I didn't see noticeable code duplication.

**Pull request merged**